tensorflow/core/kernels/sparse_to_dense_op.cc

```diff
@@ -165,4 +190,91 @@ REGISTER_KERNELS_ALL(tstring);
 #undef REGISTER_KERNELS_ALL
 #undef REGISTER_KERNELS
 
+#if GOOGLE_CUDA
+template <typename T, typename Index>
+class SparseToDenseGPU : public AsyncOpKernel {
+ public:
+  explicit SparseToDenseGPU(OpKernelConstruction* context)
+      : AsyncOpKernel(context) {
+    OP_REQUIRES_OK(context,
+                   context->GetAttr("validate_indices", &validate_indices_));
+  }
+
+  void ComputeAsync(OpKernelContext* c, DoneCallback done) final {
+    auto* stream = c->op_device_context()->stream();
+    OP_REQUIRES_ASYNC(c, stream, errors::Internal("No GPU stream available."),
+                      done);
+
+    const Tensor& indices = c->input(0);
+    const Tensor& output_shape = c->input(1);
+    const Tensor& sparse_values = c->input(2);
+    const Tensor& default_value = c->input(3);
+    OP_REQUIRES_OK_ASYNC(c, CheckSparseToDenseShapes(indices, output_shape,
+                                sparse_values, default_value), done);
+
+    const int64 num_elems = indices.dims() > 0 ? indices.dim_size(0) : 1;
+    const int64 num_dims = indices.dims() > 1 ? indices.dim_size(1) : 1;
+    const int64 num_values = sparse_values.NumElements();
+
+    auto output_shape_vec = output_shape.flat<Index>();
+    TensorShape output_tensor_shape;
+    OP_REQUIRES_OK_ASYNC(c, TensorShapeUtils::MakeShape(output_shape_vec.data(),
+                                                  output_shape_vec.size(),
+                                                  &output_tensor_shape), done);
+    Tensor* output = nullptr;
+    OP_REQUIRES_OK_ASYNC(c, c->allocate_output(0, output_tensor_shape, &output),
+                         done);
+
+    int64 dense_size = output_shape_vec.data()[0];
+    for (int i = 1; i < output_shape_vec.size(); i++) {
+      dense_size *= output_shape_vec.data()[i];
+    }
+
+    Tensor output_shape_tensor;
+    OP_REQUIRES_OK_ASYNC(c, c->allocate_temp(DataTypeToEnum<Index>::value,
+                                             {output_shape_vec.size()},
+                                             &output_shape_tensor), done);
+    auto output_shape_ptr = AsDeviceMemory(
+        output_shape_tensor.template flat<Index>().data(),
+        output_shape_tensor.template flat<Index>().size());
+    OP_REQUIRES_ASYNC(c, stream->ThenMemcpy(&output_shape_ptr,
+                                            output_shape_vec.data(),
+                                            num_dims * sizeof(Index)).ok(),
+                      errors::InvalidArgument(
+                          "failed to copy memory from host to device in "
```

Comment:
Nit: maybe make the error message more specific (what were you copying from host to device?)

Reply:
Done.